### PR TITLE
feat(serializer): Use placeholder classnames

### LINKED ISSUE #11: Use placeholder classnames
See [this](https://github.com/styled-components/jest-styled-components/pull/31). Doing this could significantly improve code reviews and avoid this problem:

screen shot 2017-06-26 at 4 20 55 pm

Notice that none of those rules changed. So it's just noise in the diff. Would be much nicer to only see the diff where it changed.

Though, that could present other problems. Like what happens when I add another styled component into the mix? That would bump all of the placeholders up. Would that be all that bad though 🤔. Perhaps @MicheleBertoli can share some thoughts

### PR DESCRIPTION
**Pull request by jhurley23**
**What**:
This replaces the glamor generated classnames in the output with placeholder "c*" classnames.  This improves diffs as they are not polluted by the change in classnames.

**Why**:
As raised in issue #11 

**How**:
Implemented a `replaceClassNames()` function taking a lot of similarity from [jest-styled-components](https://github.com/styled-components/jest-styled-components/blob/de7e6a28e5e4053e61b1d5b09f006e73dd2ef1d3/src/serializers/styleSheetSerializer.js#L36-L48)

**Other**
I've created this PR with the intention of updating it further.  I'd love to get your feedback on the best way to implement some tests on `serializer.js`.  In particular the new `replaceClassNames()` function.  
Also I'm interested to see if you have any insights on a better way to implement the regex escaping.

Are there any edge-cases that need to be considered here?

**Pic**
Please see a screenshot of it in action below:
jest-glamor

Closes #11

### CODE REVIEW

File: src/serializer.js [hunk 1]
```diff
@@ -14,12 +14,23 @@ function createSerializer(styleSheet) {
...
+      return replaceClassNames(selectors, styles, printedVal)
     } else {
       return printedVal
     }
   }
 
+  function replaceClassNames(selectors, styles, code) {
+    let index = 0
+    return selectors.reduce((acc, className) => {
+      const escapedRegex = new RegExp(
```
Comment (kentcdodds) [inline src/serializer.js]: Why don't we put this logic in its own module and test it separately. That way we can get high confidence it works for the cases we need without writing a lot of confusing tests for the serializer.

Comment (jhurley23) [inline src/serializer.js]: Would this be best within a `replaceClassNames.js` or as part of a `utils.js` (or other)?

Comment (kentcdodds) [inline src/serializer.js]: I'd put it in `replace-class-names.js` (I prefer to avoid the file casing inconsistencies of operating systems by using kebab casing for all filenames).

**Comment by jhurley23** [comment]
👍  
This also removes a lot of the extra logic from the matcher, including some of the issues you outlined in the comments regarding the `containsGlamorClassName`.

File: src/matcher.js [hunk 2]
```diff
@@ -7,48 +7,11 @@ const isAddition = line => /^\+/.test(line)
 
 const isDeletion = line => /^-/.test(line)
 
-const removeGlamorClassNames = line =>
```
Comment (kentcdodds) [inline src/matcher.js]: I'm so glad to see this gone. It was so hacky!

**Comment by kentcdodds** [comment]
I've changed the destination of this PR to be `next` so it wont be released right away. We're going to do a breaking change of glamorous soonish and I'd like to do these at the same time. That way people only have to update their snapshots once :smile:

Thanks again for all the help!

**Comment by kentcdodds** [comment]
Oh, it looks like the snapshots haven't been updated. Could you update them by running: `npm start "test --updateSnapshot"`

Then commit the files that are changed. Thanks!

**Comment by kentcdodds** [comment]
I just realized, maybe it would be better to not use the letter c as the prefix for the classes but instead use the word glamor as a prefix. That might make it a little more straightforward for people who aren't familiar with the code. What do you think?

**Comment by jhurley23** [comment]
I was thinking the same, it can even be a descriptive string.  It won't be distracting for people who understand what's going on, but for newcomers they can be informed that this is a jest-glamor-react generated string.  i.e "gjr-generated-classname-0" (although that seems too verbose)

I've also been brainstorming if there's a way we can avoid the index incrementing issue when more glamor components are added to a snapshot, but that may need more thought/time.  There's probably some solutions that work "most" of the time, or more than often but struggling to come up with a bullet-proof/100% method.

p.s  Really enjoying using glamorous and jest together, am following along closely with any glamorous issues that are upcoming.  Hoping to get some writing down sometime soon (in ref to [this](https://twitter.com/kentcdodds/status/884092780884013056) )

**Comment by kentcdodds** [comment]
Why don't we update this PR to be simply: `glamor-0` Don't want it to be too verbose as you said and I think that's fairly clear (using the word `classname` is slightly redundant). "This is a class name and it's the 0th glamor one"

**Comment by kentcdodds** [comment]
P.S.

> p.s Really enjoying using glamorous and jest together, am following along closely with any glamorous issues that are upcoming. Hoping to get some writing down sometime soon (in ref to [this](https://twitter.com/kentcdodds/status/884092780884013056) )

that's awesome! I can't wait to see what you come up with :smile:

File: src/__snapshots__/custom-sheet.test.js.snap [hunk 3]
```diff
@@ -1,81 +1,81 @@
...
 }
 
 <Wrapper>
   <section
-    className="p-4em b-papayawhip"
+    className="glamor-3 glamor-4"
   >
     <Title>
       <h1
-        className="fs-1p5em text-align-center c-palevioletred"
```
Comment (kentcdodds) [inline src/__snapshots__/custom-sheet.test.js.snap]: Ah, here's a problem I hadn't noticed before... We should only apply this to classes that are generated by `glamor`. So it must begin with `css-`. This is because:

1. You can provide glamor class names together with non-glamor class names
2. You can use this jest serializer with anything that implements a StyleSheet the way that glamor does (like cxs)

Comment (jhurley23) [inline src/__snapshots__/custom-sheet.test.js.snap]: Ah right, of course.
So just to confirm if we're given this:
```className="fs-1p5em text-align-center c-palevioletred css-12345"```
we should return
```className="fs-1p5em text-align-center c-palevioletred glamor-0"```

Comment (kentcdodds) [inline src/__snapshots__/custom-sheet.test.js.snap]: Correct! :+1:

**Comment by kentcdodds** [comment]
Wonderful! Thank you!
